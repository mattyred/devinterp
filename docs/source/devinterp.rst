devinterp package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   devinterp.backends
   devinterp.mechinterp
   devinterp.optim
   devinterp.slt

Submodules
----------

devinterp.test\_utils module
----------------------------

.. automodule:: devinterp.test_utils
   :members:
   :undoc-members:
   :show-inheritance:

devinterp.utils module
----------------------

.. automodule:: devinterp.utils
   :members:
   :undoc-members:
   :show-inheritance:

devinterp.vis\_utils module
---------------------------

.. automodule:: devinterp.vis_utils
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: devinterp
   :members:
   :undoc-members:
   :show-inheritance:
