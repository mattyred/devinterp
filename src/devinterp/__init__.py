__all__ = ["devinterp.slt", "devinterp.optim"]
