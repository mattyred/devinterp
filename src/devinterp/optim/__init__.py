from devinterp.optim.sgld import SGLD as SGLD
from devinterp.optim.sgmcmc import SGMCMC as SGMCMC
from devinterp.optim.sgmcmc import SamplingMethodLiteral as SamplingMethodLiteral
from devinterp.optim.sgnht import SGNHT as SGNHT
