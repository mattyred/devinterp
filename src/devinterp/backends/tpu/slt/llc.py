from typing import Union

import numpy as np
import torch
import torch_xla.core.xla_model as xm
from devinterp.slt.callback import SamplerCallback


class LLCEstimator(SamplerCallback):
    r"""
    Callback for estimating the Local Learning Coefficient (LLC) in a rolling fashion during a sampling process.
    It calculates the LLC based on the average loss across draws for each chain:

    $$LLC = \textrm{T} * (\textrm{avg_loss} - \textrm{init_loss})$$

    For use with :func:`devinterp.slt.sampler.sample`.

    Note:
        - `init_loss` gets set inside :func:`devinterp.slt.sample()`. It can be passed as an argument to that function,
        and if not passed will be the average loss of the supplied model over `num_chains` batches.

    :param num_draws: Number of samples to draw (should be identical to :python:`num_draws` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_draws: int
    :param num_chains: Number of chains to run (should be identical to :python:`num_chains` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_chains: int
    :param nbeta: nbeta, float (default: 1., set by sample() to utils.optimal_nbeta(dataloader)=len(batch_size)/np.log(len(batch_size)))
    :type nbeta: int
    :param device: Device to perform computations on, e.g., 'cpu' or 'cuda'.
    :type device: str | torch.device, optional
    """

    def __init__(
        self,
        num_chains: int,
        num_draws: int,
        nbeta: float,
        init_loss: torch.Tensor = None,
        device: Union[torch.device, str] = "cpu",
        eval_field: str = "loss",
    ):
        self.num_chains = num_chains
        self.num_draws = num_draws
        self.loss = torch.zeros(1, dtype=torch.float32).to(device)
        self.sq_loss = torch.zeros(1, dtype=torch.float32).to(device)
        self.init_loss = init_loss or torch.zeros(1, dtype=torch.float32).to(device)
        self.nbeta = torch.tensor(nbeta, dtype=torch.float32).to(device)
        self.device = device
        self.count = 0.0
        self.eval_field = eval_field

    def update(self, chain: int, draw: int, loss: torch.Tensor):
        self.loss += loss
        self.sq_loss += loss**2
        self.count += 1.0

    def finalize(self):
        scale = 1.0 / (self.count * xm.xrt_world_size())
        self.loss = xm.all_reduce(xm.REDUCE_SUM, self.loss, scale=scale)
        self.sq_loss = xm.all_reduce(xm.REDUCE_SUM, self.sq_loss, scale=scale)

        self.llc = self.nbeta * (self.loss - self.init_loss)

    def sample(self):
        """
        :returns: A dict :python:`{"llc/mean": llc_mean, "llc/std": llc_std, "llc-chain/{i}": llc_trace_per_chain, "loss/trace": loss_trace_per_chain}`. (Only after running :python:`devinterp.slt.sampler.sample(..., [llc_estimator_instance], ...)`).
        """
        return {
            "init_loss": self.init_loss.cpu().item(),
            "llc": self.llc.cpu().item(),
            "loss": self.loss.cpu().item(),
            "sq_loss": self.sq_loss.cpu().item(),
        }

    def __call__(self, chain: int, draw: int, *args, **kwargs):
        self.update(chain, draw, kwargs[self.eval_field])



class OnlineLLCEstimator(SamplerCallback):
    """
    Callback for estimating the Local Learning Coefficient (LLC) in an online fashion during a sampling process.
    It calculates LLCs using the same formula as :func:`devinterp.slt.llc.LLCEstimator`, but continuously and including means and std across draws (as opposed to just across chains).
    For use with :func:`devinterp.slt.sampler.sample`.

    Note:
        - `init_loss` gets set inside :func:`devinterp.slt.sample()`. It can be passed as an argument to that function,
        and if not passed will be the average loss of the supplied model over `num_chains` batches.
    :param num_draws: Number of samples to draw (should be identical to :python:`num_draws` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_draws: int
    :param num_chains: Number of chains to run (should be identical to :python:`num_chains` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_chains: int
    :param nbeta: nbeta, float (default: 1., set by sample() to utils.optimal_nbeta(dataloader)=len(batch_size)/np.log(len(batch_size)))
    :type nbeta: int
    :param device: Device to perform computations on, e.g., 'cpu' or 'cuda'. Default is 'cpu'
    :type device: str | torch.device, optional
    """

    def __init__(
        self,
        num_chains: int,
        num_draws: int,
        nbeta: float,
        init_loss,
        device="cpu",
        eval_field="loss",
    ):
        self.num_chains = num_chains
        self.num_draws = num_draws
        self.init_loss = init_loss  # gets set by devinterp.slt.sample()

        self.losses = [[] for _ in range(num_chains)]
        self.llcs = np.zeros((num_chains, num_draws))
        self.nbeta = nbeta
        self.device = device
        self.eval_field = eval_field

    def update(self, chain: int, draw: int, loss: float):
        self.losses[chain].append(loss)

    def finalize(self):
        self.losses = torch.tensor(self.losses).cpu().numpy()
        cum_loss_avgs = np.cumsum(np.mean(self.losses, axis=0), axis=0) / np.arange(
            1, self.num_draws + 1
        )
        self.llcs = self.nbeta * (cum_loss_avgs - self.init_loss)

    def sample(self):
        """
        :returns: A dict :python:`{"llc/means": llc_means, "llc/stds": llc_stds, "llc/trace": llc_trace_per_chain, "loss/trace": loss_trace_per_chain}`. (Only after running :python:`devinterp.slt.sampler.sample(..., [llc_estimator_instance], ...)`).
        """
        return {
            "init_loss": self.init_loss,
            "llc": self.llcs[-1],
            "loss": self.losses[-1],
            "llcs": self.llcs,
            "losses": self.losses,
        }

    def __call__(self, chain: int, draw: int, *args, **kwargs):
        self.update(chain, draw, kwargs[self.eval_field])


def llc_estimator_factory(
    num_chains: int,
    num_draws: int,
    nbeta: float,
    init_loss,
    device="cpu",
    eval_field: str = "loss",
    online: bool = False,
    per_token: bool = False,
) -> SamplerCallback:
    if online and per_token:
        raise ValueError("Cannot have both online and per_token set to True")
    elif online:
        return OnlineLLCEstimator(
            num_chains=num_chains,
            num_draws=num_draws,
            nbeta=nbeta,
            init_loss=init_loss,
            device=device,
            eval_field=eval_field,
        )
    # elif per_token:
    #     return PerTokenLLCEstimator(
    #         num_chains=num_chains,
    #         num_draws=num_draws,
    #         nbeta=nbeta,
    #         init_loss=init_loss,
    #         device=device,
    #         eval_field=eval_field,
    #     )
    else:
        return LLCEstimator(
            num_chains=num_chains,
            num_draws=num_draws,
            nbeta=nbeta,
            device=device,
            init_loss=init_loss,
            eval_field=eval_field,
        )
