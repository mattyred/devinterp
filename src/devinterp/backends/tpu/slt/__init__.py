from devinterp.backends.tpu.slt.llc import *
from devinterp.backends.tpu.slt.sampler import *
