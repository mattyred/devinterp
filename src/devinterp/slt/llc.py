from typing import Union, Optional
import torch
from devinterp.slt.callback import SamplerCallback
import warnings

# TPU support
try:
    import torch_xla.core.xla_model as xm
    USING_XLA = True
except ImportError:
    USING_XLA = False

class LLCEstimator(SamplerCallback):
    r"""
    Callback for estimating the Local Learning Coefficient (LLC) in a rolling fashion during a sampling process.
    It calculates the LLC based on the average loss across draws for each chain:
    
    $$LLC = \textrm{T} * (\textrm{avg_loss} - \textrm{init_loss})$$

    For use with :func:`devinterp.slt.sampler.sample`.
    
    Note: 
        - `init_loss` gets set inside :func:`devinterp.slt.get_results()`. It can be passed as an argument to that function, 
        and if not passed will be the average loss of the supplied model over `num_chains` batches.

    :param num_draws: Number of samples to draw (should be identical to :python:`num_draws` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_draws: int
    :param num_chains: Number of chains to run (should be identical to :python:`num_chains` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_chains: int
    :param nbeta: Reparameterized inverse temperature
    :type nbeta: float
    :param device: Device to perform computations on, e.g., 'cpu' or 'cuda'.
    :type device: str | torch.device, optional
    """
    def __init__(
        self,
        num_chains: int,
        num_draws: int,
        init_loss: torch.Tensor,
        eval_field: str = "loss",
        device: Union[torch.device, str] = "cpu",
        nbeta: float = None,
        temperature: Optional[float] = None, # Temperature is deprecated
    ):
        self.num_chains = num_chains
        self.num_draws = num_draws

        self.loss = torch.zeros((num_chains, num_draws), dtype=torch.float32).to(device)
        self.init_loss = init_loss

        assert nbeta is not None or temperature is not None, "Please provide a value for nbeta."
        if nbeta is None and temperature is not None:
            nbeta = temperature
            warnings.warn("Temperature is deprecated. Please use nbeta instead.")
        self.nbeta = torch.tensor(nbeta, dtype=torch.float32).to(device)
        self.eval_field = eval_field
        self.device = device
        self.count = 0.0

    def update(self, chain: int, draw: int, loss: torch.Tensor, **kwargs):
        """
        Called at each step of sampling.

        Order of operations: 
        1. update() is called at each draw.
        2. finalize() is called at the end of sampling.
        3. get_results() is called after finalize().

        :param draw: The current draw index.
        :type draw: int
        :param loss: The loss at the current draw.
        :type loss: float
        :param chain: The chain index. Note that this is not used here at all.
        :type chain: int
        """
        self.loss[chain, draw] = loss
        self.count += 1.0

    def finalize(self):
        """Called once at the end of sampling."""

        # If using TPUs, aggregate the loss across all devices.
        if USING_XLA:
            if self.count == 0:
                self.count = 1
            scale = 1.0 / (self.count * xm.xrt_world_size())
            self.loss = xm.all_reduce(xm.REDUCE_SUM, self.loss, scale=scale)

        avg_loss = self.loss.mean(axis=1) # [num_chains]. Average loss across draws for each chain.

        self.llc_per_chain = self.nbeta * (avg_loss - self.init_loss) # [num_chains]. A single LLC value for each chain.
        self.sq_loss = torch.square(self.loss)
        self.llc_mean = self.llc_per_chain.mean() # Mean LLC across chains.
        self.llc_std = self.llc_per_chain.std() # Standard deviation of LLC across chains.

    def get_results(self):
        """
        Method to return results. Called after finalize().

        :returns: A dict :python:`{"llc/mean": llc_mean, "llc/std": llc_std, "llc-chain/{i}": llc_trace_per_chain, "loss/trace": loss_trace_per_chain}`. (Only after running :python:`devinterp.slt.sampler.sample(..., [llc_estimator_instance], ...)`).
        """
        return {
            "llc/mean": self.llc_mean.cpu().numpy().item(),
            "llc/std": self.llc_std.cpu().numpy().item(),
            **{
                f"llc-chain/{i}": self.llc_per_chain[i].cpu().numpy().item()
                for i in range(self.num_chains)
            },
            "loss/trace": self.loss.cpu().numpy(),
            "loss/init_loss": self.init_loss,
            "loss/sq_loss": self.sq_loss.cpu().numpy(),
        }

    def __call__(self, chain: int, draw: int, **kwargs):
        """Calls self.update."""
        self.update(chain, draw, kwargs[self.eval_field])


class OnlineLLCEstimator(SamplerCallback):
    """
    Callback for estimating the Local Learning Coefficient (LLC) in an online fashion during a sampling process.
    It calculates LLCs using the same formula as :func:`devinterp.slt.llc.LLCEstimator`, but continuously and including means and std across draws (as opposed to just across chains).
    For use with :func:`devinterp.slt.sampler.sample`.
    
    Note: 
        - `init_loss` gets set inside :func:`devinterp.slt.get_results()`. It should be passed as an argument to that function.
    :param num_draws: Number of samples to draw (should be identical to :python:`num_draws` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_draws: int
    :param num_chains: Number of chains to run (should be identical to :python:`num_chains` passed to :python:`devinterp.slt.sampler.sample`)
    :type num_chains: int
    :param nbeta: Inverse reparameterized temperature
    :type nbeta: float
    :param device: Device to perform computations on, e.g., 'cpu' or 'cuda'. Default is 'cpu'
    :type device: str | torch.device, optional
    """

    def __init__(
        self, 
        num_chains: int, 
        num_draws: int, 
        init_loss: torch.Tensor,
        device="cpu",
        eval_field="loss",
        nbeta: float = None,
        temperature: Optional[float] = None, # Temperature is deprecated
    ):
        self.num_draws = num_draws
        self.init_loss = init_loss

        self.losses = torch.zeros((num_chains, num_draws)).to(device)
        self.llcs = torch.zeros((num_chains, num_draws)).to(device)

        assert nbeta is not None or temperature is not None, "Please provide a value for nbeta."
        if nbeta is None and temperature is not None:
            nbeta = temperature
            warnings.warn("Temperature is deprecated. Please use nbeta instead.")
        self.nbeta = torch.tensor(nbeta, dtype=torch.float32).to(device)
        
        self.device = device
        self.eval_field = eval_field

        self.llc_means = torch.tensor(num_draws, dtype=torch.float32).to(device)
        self.llc_stds = torch.tensor(num_draws, dtype=torch.float32).to(device)
        self.moving_avg_llcs = torch.zeros(
            (num_chains, num_draws), dtype=torch.float32
        ).to(device)
        
    def update(self, chain: int, draw: int, loss: float):
        self.losses[chain, draw] = loss
        self.llcs[chain, draw] = self.nbeta * (loss - self.init_loss)
        if draw == 0:
            self.moving_avg_llcs[chain, draw] = self.nbeta * (
                loss - self.init_loss
            )
        else:
            t = draw + 1
            prev_moving_avg = self.moving_avg_llcs[chain, draw - 1]
            self.moving_avg_llcs[chain, draw] = (1 / t) * (
                (t - 1) * prev_moving_avg + self.nbeta * (loss - self.init_loss)
            )

    def finalize(self):
        self.llc_means = self.llcs.mean(dim=0)
        self.llc_stds = self.llcs.std(dim=0)

    def get_results(self):
        """    
        :returns: A dict :python:`{"llc/means": llc_means, "llc/stds": llc_stds, "llc/trace": llc_trace_per_chain, "loss/trace": loss_trace_per_chain}`. (Only after running :python:`devinterp.slt.sampler.sample(..., [llc_estimator_instance], ...)`).
        """
        return {
            "llc/means": self.llc_means.cpu().numpy(),
            "llc/moving_avg": self.moving_avg_llcs.cpu().numpy(),
            "llc/stds": self.llc_stds.cpu().numpy(),
            "llc/trace": self.llcs.cpu().numpy(),
            "loss/trace": self.losses.cpu().numpy(),
        }

    def __call__(self, chain: int, draw: int, **kwargs):
        self.update(chain, draw, kwargs[self.eval_field])
