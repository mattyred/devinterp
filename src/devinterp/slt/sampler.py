import itertools
import warnings
from copy import deepcopy
from typing import Callable, Dict, List, Literal, Optional, Type, Union

import torch
from torch import nn
from torch.multiprocessing import cpu_count, get_context
from torch.utils.data import DataLoader
from tqdm import tqdm

from devinterp.backends.default.slt.llc import LLCEstimator, OnlineLLCEstimator
from devinterp.backends.default.slt.mala import MalaAcceptanceRate
from devinterp.backends.default.slt.norms import NoiseNorm
from devinterp.optim.sgld import SGLD
from devinterp.slt.callback import SamplerCallback, validate_callbacks
from devinterp.utils import (
    USE_TPU_BACKEND,
    EvaluateFn,
    call_with,
    get_init_loss_multi_batch,
    optimal_temperature,
    prepare_input,
    split_results,
)

if USE_TPU_BACKEND:
    from devinterp.backends.tpu.slt.sampler import sample
else:
    from devinterp.backends.default.slt.sampler import sample


def estimate_learning_coeff_with_summary(
    model: torch.nn.Module,
    loader: DataLoader,
    callbacks: List[Callable] = [],
    evaluate: Optional[EvaluateFn] = None,
    sampling_method: Type[torch.optim.Optimizer] = SGLD,
    optimizer_kwargs: Optional[Dict[str, Union[float, Literal["adaptive"]]]] = None,
    num_draws: int = 100,
    num_chains: int = 10,
    num_burnin_steps: int = 0,
    num_steps_bw_draws: int = 1,
    init_loss: float = None,
    grad_accum_steps: int = 1,
    cores: int = 1,
    seed: Optional[Union[int, List[int]]] = None,
    device: Union[torch.device, str] = torch.device("cpu"),
    verbose: bool = True,
    optimize_over_per_model_param: Optional[Dict[str, List[bool]]] = None,
    online: bool = False,
) -> dict:
    optimizer_kwargs.setdefault("temperature", optimal_temperature(loader))
    if not init_loss:
        init_loss = get_init_loss_multi_batch(
            loader, num_chains, model, evaluate, device
        )
        # alternative: init_loss = get_init_loss_full_batch(loader, model, evaluate, device)
        # alternative: init_loss = get_init_loss_one_batch(loader, model, evaluate, device)
    if online:
        llc_estimator = OnlineLLCEstimator(
            num_chains, num_draws, optimizer_kwargs["temperature"], device=device
        )
    else:
        llc_estimator = LLCEstimator(
            num_chains, num_draws, optimizer_kwargs["temperature"], device=device
        )

    callbacks = [llc_estimator, *callbacks]

    sample(
        model=model,
        loader=loader,
        evaluate=evaluate,
        sampling_method=sampling_method,
        optimizer_kwargs=optimizer_kwargs,
        num_draws=num_draws,
        num_chains=num_chains,
        num_burnin_steps=num_burnin_steps,
        num_steps_bw_draws=num_steps_bw_draws,
        grad_accum_steps=grad_accum_steps,
        cores=cores,
        seed=seed,
        device=device,
        verbose=verbose,
        callbacks=callbacks,
        init_loss=init_loss,
        optimize_over_per_model_param=optimize_over_per_model_param,
    )

    results = {}

    for callback in callbacks:
        if hasattr(callback, "sample"):
            results.update(callback.sample())

    return results


def estimate_learning_coeff(
    model: torch.nn.Module,
    loader: DataLoader,
    callbacks: List[Callable] = [],
    evaluate: Optional[EvaluateFn] = None,
    sampling_method: Type[torch.optim.Optimizer] = SGLD,
    optimizer_kwargs: Optional[Dict[str, Union[float, Literal["adaptive"]]]] = None,
    num_draws: int = 100,
    num_chains: int = 10,
    num_burnin_steps: int = 0,
    num_steps_bw_draws: int = 1,
    init_loss: float = None,
    grad_accum_steps: int = 1,
    cores: int = 1,
    seed: Optional[Union[int, List[int]]] = None,
    device: Union[torch.device, str] = torch.device("cpu"),
    verbose: bool = True,
    optimize_over_per_model_param: Optional[Dict[str, List[bool]]] = None
) -> float:
    return estimate_learning_coeff_with_summary(
        model=model,
        loader=loader,
        evaluate=evaluate,
        sampling_method=sampling_method,
        optimizer_kwargs=optimizer_kwargs,
        num_draws=num_draws,
        num_chains=num_chains,
        num_burnin_steps=num_burnin_steps,
        num_steps_bw_draws=num_steps_bw_draws,
        grad_accum_steps=grad_accum_steps,
        cores=cores,
        seed=seed,
        device=device,
        verbose=verbose,
        callbacks=callbacks,
        online=False,
        init_loss=init_loss,
        optimize_over_per_model_param=optimize_over_per_model_param,
    )["llc/mean"]
