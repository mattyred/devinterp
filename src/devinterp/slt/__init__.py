from devinterp.slt.sampler import *