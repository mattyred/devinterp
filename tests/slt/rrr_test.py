import numpy as np
import pytest
import torch
import torch.nn.functional as F
import platform
from devinterp.backends.default.slt.sampler import sample
from devinterp.optim import SGLD, SGMCMC
from devinterp.slt.llc import LLCEstimator
from devinterp.utils import default_nbeta, evaluate_mse, get_init_loss_multi_batch
from torch.utils.data import DataLoader, TensorDataset

# Test configuration constants
TOLERANCE_RTOL = 0.4
CONFIDENCE_MULTIPLIER = 2.5
FULL_SAMPLING_DRAWS = 400
SNAPSHOT_DRAWS = 5
LEARNING_RATE = 0.0006
LOCALIZATION = 1.0
NUM_CHAINS = 3
RANDOM_SEED = 42
NUM_SAMPLES = 1000


# not a fixture as we're generating data for several m, n combinations
# and I couldn't figure out how to fit that into the fixture mold
def generated_rrr_dataset(m, n):
    """Generate synthetic dataset for reduced rank regression testing.

    Args:
        m: Input dimension
        n: Output dimension

    Returns:
        Tuple of (dataloader, dataset, input_tensor, output_tensor)
    """
    torch.manual_seed(RANDOM_SEED)
    np.random.seed(RANDOM_SEED)
    x = torch.randn(NUM_SAMPLES, m)
    y = torch.randn(NUM_SAMPLES, n)
    train_data = TensorDataset(x, y)

    # Add deterministic generator
    generator = torch.Generator()
    generator.manual_seed(RANDOM_SEED)
    train_dataloader = DataLoader(train_data, batch_size=NUM_SAMPLES, shuffle=True)
    return train_dataloader, train_data, x, y


# Test case mapping for theoretical scenarios from Aoyagi & Watanabe (2004)
TEST_CASES = [
    (5, 3, 5, "case_1_odd", "General case with odd sum of dimensions"),
    (5, 4, 5, "case_1_even", "General case with even sum of dimensions"),
    (4, 3, 8, "case_2", "Input + hidden < output dimension"),
    (8, 3, 4, "case_3", "Output + hidden < input dimension"),
    (3, 8, 4, "case_4", "Input + output < hidden dimension"),
]


@pytest.mark.skipif(
    platform.machine() != "x86_64",
    reason=f"Differences in results between ARM and x86_64. Your arch is {platform.machine()}",
)
@pytest.mark.parametrize("sampling_method", [SGLD, SGMCMC.sgld])
@pytest.mark.parametrize(
    "m,h,n,case_name,description",
    TEST_CASES,
    ids=[f"{case[3]}_{case[0]}x{case[1]}x{case[2]}" for case in TEST_CASES],
)
@pytest.mark.parametrize("perturb", [True, False], ids=["perturbed", "unperturbed"])
def test_accuracy_rrr(
    sampling_method,
    m,
    h,
    n,
    case_name,
    description,
    perturb,
    ReducedRankRegressor,
    snapshot,
    is_snapshot_update,
):
    """Test reduced rank regression LLC estimation accuracy.

    Tests the Local Learning Coefficient (LLC) estimation for different
    dimensional configurations of reduced rank regression models.

    Based on "The Generalization Error of Reduced Rank Regression in Bayesian Estimation",
    M. Aoyagi & S. Watanabe, 2004.

    Args:
        sampling_method: SGLD or SGMCMC sampling method
        m: Input dimension
        h: Hidden/rank dimension
        n: Output dimension
        case_name: Theoretical case identifier
        description: Human-readable description of the test case
        perturb: Whether to perturb the model away from optimal parameters
        ReducedRankRegressor: Model class fixture
        snapshot: Expected LLC value for regression testing
        is_snapshot_update: Whether to run full accuracy test against theory
    """
    # Set up test data and model
    model, train_dataloader = _setup_rrr_model(m, h, n, perturb, ReducedRankRegressor)

    # Run the snapshot test first, so our random seed is consistent.
    llc_mean, llc_std_dev = do_sampling(
        sampling_method, train_dataloader, model, num_draws=SNAPSHOT_DRAWS
    )

    # Test against theoretical values when updating snapshots
    if is_snapshot_update:
        # Verify that the calculated case matches the expected case from pytest parameters
        calculated_case, _ = calc_true_lc(m, h, n)
        assert calculated_case == case_name, (
            f"Calculated theoretical case '{calculated_case}' does not match expected case '{case_name}' "
            f"for dimensions (M={m}, H={h}, N={n})"
        )

        _test_theoretical_accuracy(
            sampling_method, train_dataloader, model, m, h, n, perturb
        )

    # Always test against snapshot for consistency.
    assert llc_mean == snapshot


def _setup_rrr_model(m, h, n, perturb, ReducedRankRegressor):
    """Set up the reduced rank regression model for testing."""
    torch.manual_seed(RANDOM_SEED)
    np.random.seed(RANDOM_SEED)

    criterion = F.mse_loss
    train_dataloader, train_data, x, y = generated_rrr_dataset(m, n)
    # Update dimensions based on actual data
    m = x.size(1)
    n = y.size(1)

    model = ReducedRankRegressor(m, h, n, x, y, criterion)

    if perturb:
        # We repeat the Litany Against Non-Determinism:
        torch.manual_seed(RANDOM_SEED)
        np.random.seed(RANDOM_SEED)
        model.perturb()

    return model, train_dataloader


def _test_theoretical_accuracy(
    sampling_method, train_dataloader, model, m, h, n, perturb
):
    """Test LLC estimation against theoretical values with full sampling."""
    llc_mean, llc_std_dev = do_sampling(
        sampling_method, train_dataloader, model, num_draws=FULL_SAMPLING_DRAWS
    )
    case, true_lc = calc_true_lc(m, h, n)

    if not perturb:
        _assert_close_to_theory(
            llc_mean, llc_std_dev, true_lc, case, m, h, n, sampling_method
        )
    else:
        _assert_not_close_to_theory(
            llc_mean, llc_std_dev, true_lc, case, m, h, n, sampling_method
        )


def _assert_close_to_theory(
    llc_mean, llc_std_dev, true_lc, case, m, h, n, sampling_method
):
    """Assert that LLC estimate is close to theoretical value."""
    confidence_interval = CONFIDENCE_MULTIPLIER * llc_std_dev
    error_msg = (
        f"DLN case {case}: LLC estimate {llc_mean:.3f} ± {confidence_interval:.3f} "
        f"should be close to theoretical LC {true_lc:.3f} "
        f"for dimensions (M={m}, H={h}, N={n}) using {sampling_method}"
    )
    assert np.isclose(llc_mean, true_lc, rtol=TOLERANCE_RTOL), error_msg


def _assert_not_close_to_theory(
    llc_mean, llc_std_dev, true_lc, case, m, h, n, sampling_method
):
    """Assert that perturbed model LLC estimate is not close to theoretical value."""
    confidence_interval = CONFIDENCE_MULTIPLIER * llc_std_dev
    error_msg = (
        f"Perturbed model should not match theory. "
        f"DLN case {case}: LLC estimate {llc_mean:.3f} ± {confidence_interval:.3f} "
        f"vs theoretical LC {true_lc:.3f} "
        f"for dimensions (M={m}, H={h}, N={n}) using {sampling_method}"
    )
    assert not np.isclose(llc_mean, true_lc, rtol=TOLERANCE_RTOL), error_msg


def do_sampling(sampling_method, train_dataloader, model, num_draws):
    """Perform MCMC sampling to estimate LLC.

    Args:
        sampling_method: SGLD or SGMCMC sampling method
        train_dataloader: DataLoader for training data
        model: Model to sample from
        num_draws: Number of MCMC draws to perform

    Returns:
        Tuple of (llc_mean, llc_std_dev)
    """
    torch.manual_seed(RANDOM_SEED)
    np.random.seed(RANDOM_SEED)

    init_loss = get_init_loss_multi_batch(
        train_dataloader, NUM_CHAINS, model, evaluate_mse, device="cpu"
    )

    llc_estimator = LLCEstimator(
        num_chains=NUM_CHAINS,
        num_draws=num_draws,
        nbeta=default_nbeta(train_dataloader),
        init_loss=init_loss,
    )

    sample(
        model,
        train_dataloader,
        evaluate=evaluate_mse,
        sampling_method_kwargs=dict(
            lr=LEARNING_RATE,
            localization=LOCALIZATION,
            nbeta=default_nbeta(train_dataloader),
        ),
        sampling_method=sampling_method,
        num_chains=NUM_CHAINS,
        num_draws=num_draws,
        callbacks=[llc_estimator],
        verbose=False,
        seed=RANDOM_SEED,
    )

    results = llc_estimator.get_results()
    return results["llc/mean"], results["llc/std"]


def calc_true_lc(m, h, n):
    """Calculate theoretical learning coefficient for reduced rank regression.

    Based on the theoretical results from Aoyagi & Watanabe (2004).

    Args:
        m: Input dimension
        h: Hidden/rank dimension
        n: Output dimension

    Returns:
        Tuple of (case_name, theoretical_lc_value)
    """
    # Determine which theoretical case applies
    case_2 = m + h < n
    case_3 = n + h < m
    case_4 = m + n < h
    case_1_even = not (case_2 or case_3 or case_4) and (m + h + n) % 2 == 0
    case_1_odd = not (case_2 or case_3 or case_4) and (m + h + n) % 2 == 1

    if case_2:
        return "case_2", m * h / 2
    elif case_3:
        return "case_3", h * n / 2
    elif case_4:
        return "case_4", m * n / 2
    elif case_1_even:
        numerator = 2 * m * n + 2 * h * n + 2 * m * h - n**2 - m**2 - h**2
        return "case_1_even", numerator / 8
    elif case_1_odd:
        numerator = 1 + 2 * m * n + 2 * h * n + 2 * m * h - n**2 - m**2 - h**2
        return "case_1_odd", numerator / 8
    else:
        raise ValueError(
            f"Unknown theoretical case for dimensions (M={m}, H={h}, N={n})"
        )


# TODO:
# Scale up these estimates like in Furman & Lau (2024), also to DLNs more generally
#
# For models with a closed-form population loss, like DLNs:
# compare SGLD on empirical loss with SGLD on population loss (results should agree)
# SGLD on population loss should be able to get the LLC exactly correct,
# assuming beta is sufficiently high (using population loss here instead of empirical loss allows very high beta without prohibitively large training set size)
