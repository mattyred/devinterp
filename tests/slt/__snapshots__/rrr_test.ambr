# serializer version: 1
# name: test_accuracy_rrr[perturbed-case_1_even_5x4x5-SGLD]
  -950.279052734375
# ---
# name: test_accuracy_rrr[perturbed-case_1_even_5x4x5-sgld]
  -950.279052734375
# ---
# name: test_accuracy_rrr[perturbed-case_1_odd_5x3x5-SGLD]
  -540.136474609375
# ---
# name: test_accuracy_rrr[perturbed-case_1_odd_5x3x5-sgld]
  -540.136474609375
# ---
# name: test_accuracy_rrr[perturbed-case_2_4x3x8-SGLD]
  -428.4161071777344
# ---
# name: test_accuracy_rrr[perturbed-case_2_4x3x8-sgld]
  -428.4161071777344
# ---
# name: test_accuracy_rrr[perturbed-case_3_8x3x4-SGLD]
  -1717.2952880859375
# ---
# name: test_accuracy_rrr[perturbed-case_3_8x3x4-sgld]
  -1717.2952880859375
# ---
# name: test_accuracy_rrr[perturbed-case_4_3x8x4-SGLD]
  -5700.07275390625
# ---
# name: test_accuracy_rrr[perturbed-case_4_3x8x4-sgld]
  -5700.07275390625
# ---
# name: test_accuracy_rrr[unperturbed-case_1_even_5x4x5-SGLD]
  0.36638423800468445
# ---
# name: test_accuracy_rrr[unperturbed-case_1_even_5x4x5-sgld]
  0.36638423800468445
# ---
# name: test_accuracy_rrr[unperturbed-case_1_odd_5x3x5-SGLD]
  0.3891811668872833
# ---
# name: test_accuracy_rrr[unperturbed-case_1_odd_5x3x5-sgld]
  0.3891811668872833
# ---
# name: test_accuracy_rrr[unperturbed-case_2_4x3x8-SGLD]
  0.13740848004817963
# ---
# name: test_accuracy_rrr[unperturbed-case_2_4x3x8-sgld]
  0.13740848004817963
# ---
# name: test_accuracy_rrr[unperturbed-case_3_8x3x4-SGLD]
  0.4877088963985443
# ---
# name: test_accuracy_rrr[unperturbed-case_3_8x3x4-sgld]
  0.4877088963985443
# ---
# name: test_accuracy_rrr[unperturbed-case_4_3x8x4-SGLD]
  0.4049715995788574
# ---
# name: test_accuracy_rrr[unperturbed-case_4_3x8x4-sgld]
  0.4049715995788574
# ---
